### ISSUE TITLE
can't reproduce the 2.723 Chamfer Distance result

### ISSUE THREAD

**Issue by AlphaPav** [hzxie/GRNet#8]
You need to train the whole network with Chamfer Distance. It reaches CD ~0.40 on ShapeNet.
Then, you need to fine-tune the network with Gridding Loss + Chamfer Distance on the Coarse Point Cloud.
Finally, you fine-tune the network with Chamfer Distance. Chamfer Distance is taken as a metric, therefore, you cannot get lower CD without using Chamfer Distance as a loss.

_Originally posted by @hzxie in https://github.com/hzxie/GRNet/issues/3#issuecomment-656446671_

**Comment by hzxie** [comment]
So, what wrong with you?

**Comment by AlphaPav** [comment]
Hi author, thanks for the amazing work.

On your released pre-trained model, I can get 0.7082 F-score 2.722 CD. 
However, when I train from scratch, I had some problems listed below: 

"You need to train the whole network with Chamfer Distance." --- It reaches 4.588 CD,  0.6133 F-score, which is similar with Table 7&Not Used&CD&Complete = 4.460 in your paper.

"Then .. fine-tune the network with Gridding Loss + Chamfer Distance on the Coarse Point Cloud." ---- It reaches 4.536 CD, 0.6255 F-score. It was supposed to be about ~2.7, right? 

"Finally, you fine-tune the network with Chamfer Distance." --- the CD didn't decrease below 4.536. 

I'm wondering what steps am I making mistakes? (like learning rate/loss weight of gridding loss)

**Comment by AlphaPav** [comment]
> So, what's the problem you are facing now?
> Please provide more details.

Hi author, do you have any ideas?

**Comment by AlphaPav** [comment]
your  [processed ShapeNet dataset ](https://gateway.infinitescript.com/?fileName=ShapeNetCompletion) has 28974 training data samples 
while [the PCN dataset](https://drive.google.com/drive/folders/1P_W1tz5Q4ZLapUifuOE4rFAZp6L1XTJz) has  231792 training data samples 


is it because your provided dataset is not completed?

**Comment by AlphaPav** [comment]
Hi,
do you have any ideas? @hzxie

**Comment by hzxie** [comment]
@AlphaPav 
Sorry for the late reply. I don't have time to check this issue these days.
But I'm sure that there is nothing wrong with the released dataset. 231792 / 28974 = 8, which indicates that there are 8 partial input point cloud for each model in ShapeNet.

**Comment by AlphaPav** [comment]
> @AlphaPav
> Sorry for the late reply. I don't have time to check this issue these days.
> But I'm sure that there is nothing wrong with the released dataset. 231792 / 28974 = 8, which indicates that there are 8 partial input point cloud for each model in ShapeNet.

The PCN dataset is about 48 GB, while the released dataset is about 10 GB. Do you mean that you randomly augment each point cloud 8 times during training?

**Comment by hzxie** [comment]
No. I think the difference may be caused by different compression ratios.
You can also generate the ShapeNet dataset from [PCN](https://drive.google.com/drive/folders/1P_W1tz5Q4ZLapUifuOE4rFAZp6L1XTJz) with [this script](https://github.com/hzxie/GRNet/blob/master/utils/lmdb_serializer.py).

**Comment by SarahChane98** [comment]
Hi! I also cannot reproduce the results. The highest CD I got after training three times was 5.2. May I know how many epochs you've trained for each round respectively? (i.e. CD only, CD + gridding loss, CD only)

**Comment by hzxie** [comment]
@SarahChane98 
I cannot report the exact numbers of epochs for each round.
For each round, I train several times until the loss does not decrease.

**Comment by paulwong16** [comment]
Hi there, I just tested your pretrained model on test set, and the result is close to the value reported in paper. However, when I tested on validation dataset, it reported a dense CD around 7.177. I was wondering why there is a hugh gap between CDs on val set and test set?

**Comment by hzxie** [comment]
@paulwong16 
Because we choose the best model for the test set instead of the validation set.

**Comment by paulwong16** [comment]
> @paulwong16 
> 
> Because we choose the best model for the test set instead of the validation set.

but why CD on test set could be even much lower than on training set?

**Comment by hzxie** [comment]
Because the pretrained model is best the fit distribution of testing set.
Instead, the distribution of the training and validation set may be different from the testing set.

**Comment by paulwong16** [comment]
> @paulwong16
> 
> Because the pretrained model is best for fitting distribution of the testing set.
> 
> Instead, the distribution of the training and validation set may be different from the testing set.

well...i believe the best model should not be chosen according to the test result (instead, should be the validation result). And from the best results I could reproduce, the training loss was a little lower than val loss and test loss, and test loss was close to the val loss. 

Anyway, thanks for your kind reply, I will try to reproduce the result.

**Comment by hzxie** [comment]
Yes, choosing models from the testing set is not a good option.
For the Completion 3D benchmark, the best model is chosen from the validation set. (Because we don't have the ground truth for the testing set)

**Comment by wangyida** [comment]
@hzxie Hi I'm wondering how you incoorporate gridding_loss in training? I have not found it in the script Thanks

**Comment by hzxie** [comment]
You can use the Gridding Loss here:

https://github.com/hzxie/GRNet/blob/335259235804fa30b0e89e0e257d886687bfb6f3/core/train.py#L113

when fine-tuning the network.

**Comment by wangyida** [comment]
@hzxie Thank you, I tried it out and the result seems to be fitting with the expected trends. Thanks for your inspiring work;)

**Comment by Lillian9707** [comment]
Hi, I'm wondering how to fine-tune the network with the previous weight? I've tried the same configuration as your paper but the best model gets CD=4.538 and F-Score=6.206 while your pre-trained model can get CD=2.723 and F-Score=7.082.

And I check the log and find that the network had converged to the optimal in 20 epochs. Why you set 150 epoch as the default?

**Comment by hzxie** [comment]
@Lillian9707 

In my experiments, the loss will continue to decrease after 20 epochs.
Moreover, you need to fine-tune the network with the Gridding Loss.

**Comment by Lillian9707** [comment]
Hi, I still cannot reproduce the result. Can you provide more details?

**Comment by hzxie** [comment]
@Lillian9707 
Keep the learning rate unchanged during fine-tuning.
According to the experimental results of AlphaPav, the CD and F-Score got better after applying Gridding Loss.

**Comment by Lillian9707** [comment]
Thank you for your reply!
But AlphaPav only gets '4.536 CD, 0.6255 F-score' after fine-tuning, which looks more stochastic.
So the fine-tuning process is to train with 1* CD on both Sparse and Dense Point Cloud + 1* Gridding Loss on Sparse Point Cloud? And the learning rate is always 5e-5?

**Comment by Lillian9707** [comment]
hi, sorry to bother you. I still cannot reproduce the results in the paper.

**Comment by hzxie** [comment]
@Lillian9707 

Try to fine-tune the network w/ and w/o Gridding Loss several times.
During fine-tuning, try to use top-10 (not always the best) weights from the previous training should be loaded.
The init learning rate for fine-tuning should be 1e-4.

**Comment by Lillian9707** [comment]
@hzxie  Hi, thank you for your kind reply! I also have a question about how you compute the metric for the coarse point cloud.
I have tried your pre-trained model for ShapeNet and change the `_metrics = Metrics.get(dense_ptcloud, data['gtcloud'])`  into `_metrics = Metrics.get(sparse_ptcloud, data['gtcloud'])` from core\test.py, but I get the CD=34.443, F-Score=0.3619 while your paper mention the CD=9.324, F-Score=0.386.
